Check assertion and assume incrementally using the push-pop functions

lucasccordeiro opened the pull request:
This PR allows us to check each assertion and assume incrementally using the push-pop functions available in the SMT solvers. If the assertion or assume statement turns to be true, we remove it from the resulting logical formula to be passed to the SMT solver; otherwise, we negate the assertion (for assertion) and add an assume(false) to cut the subsequent execution paths.

lucasccordeiro committed:
Check assertions incrementally during symex

This commit allows us to check each assertion
incrementally using the push/pop functions
available in the SMT solvers. If the assertion
turns to be true, we simplify it; othwewise,
we negate the assertion and add the resulting
expression as an assumption.

lucasccordeiro committed:
added regression tests for incremental SMT solving

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

mikhailramalho commented on src/goto-symex/symex_main.cpp:
```diff
@@ -54,6 +54,30 @@ void goto_symext::claim(const expr2tc &claim_expr, const std::string &msg)
+      if(res.is_true())
+        // we don't add this assertion to the resulting VCs
+        return;
+      else if(res.is_false())
```
No need for an `else` here.

lucasccordeiro replied:
Fixed.

mikhailramalho commented on src/goto-symex/symex_main.cpp:
```diff
@@ -54,6 +54,30 @@ void goto_symext::claim(const expr2tc &claim_expr, const std::string &msg)
+      {
+        // negate the claim and add it as an assumption
+        not2tc not_new_expr(new_expr);
+        assume(not_new_expr);
```
I don't think we need this here: we can convert it to an `assume(false)`.

Also, I'd be interested in halting the symex here: we know this is a assert violation, so we can save a lot of time by stopping and printing the cex.

lucasccordeiro committed:
simplified the code

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
added test case with a large loop bound

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
Stop symex once we find a property violation

Halt the symex once we know the assertion is
violated to save time by stopping and
printing the cex.

lucasccordeiro committed:
added test cases for incremental solving

lucasccordeiro committed:
check for verification unknown

The SMT solver can return true, false or unkown
when checking the assertions. In the unknown case,
we just report that back to the user.

rafaelsamenezes commented on src/goto-symex/symex_main.cpp:
```diff
@@ -54,6 +54,31 @@ void goto_symext::claim(const expr2tc &claim_expr, const std::string &msg)
+    }
+    catch(runtime_encoded_equationt::dual_unsat_exception &e)
+    {
+      assert(0 && "Can't check this assertion, sorry");
```
Who throws this exception? Does it give more info? If so, try to explain the reason why it was throw.

lucasccordeiro committed:
improved error message when checking the equation

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

mikhailramalho reviewed (changes requested):
This is more convoluted than it has to be, how about:
1. ask question
2. if it's true, return an do nothing
3. if it's false, add assume(false) and return
4. Otherwise, print message saying that we found a constraint that cannot be solved and we'll fall back to solve the whole equation.

Other comments:
1. No need for else if the previous if-block has a return.
2. Do we need to abort when we catch a dual_unsat_exception exception?

lucasccordeiro committed:
refactored the code

Moved the incremental check procedure to a method
to be reused by other symex functions.

lucasccordeiro committed:
Check assume statements incrementally

Check assume statements incrementally using the push/pop
functions available in the SMT solvers.

lucasccordeiro committed:
added test cases for checking assume statements

lucasccordeiro committed:
refactored the code

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

mikhailramalho commented on src/goto-symex/symex_main.cpp:
```diff
@@ -54,6 +93,13 @@ void goto_symext::claim(const expr2tc &claim_expr, const std::string &msg)
+
   cur_state->guard.guard_expr(new_expr);
   cur_state->global_guard.guard_expr(new_expr);
   remaining_claims++;
```
Could use also replace this code with a call to the new function `assertion`?

lucasccordeiro replied:
Done.

mikhailramalho commented on src/goto-symex/symex_main.cpp:
```diff
@@ -25,6 +25,45 @@
+      "another solver");
+  }
+  return false;
+}
```
add a newline here.

lucasccordeiro replied:
Fixed.

mikhailramalho commented on src/goto-symex/symex_main.cpp:
```diff
@@ -313,6 +373,12 @@ void goto_symext::symex_assume()
+    if(check_incremental(cond, false, std::string()))
+      // incremental verification has succeeded
+      return;
+  }
```
Is this correct? We should return when the early if the assume is true, but `check_incremental` also returns true when cond evaluates to false.

You'll probably need to replace the return value with a threeval object or an optional.

lucasccordeiro replied:
The method `check_incremental` already adds `assume(false)`.

mikhailramalho commented on src/goto-symex/symex_main.cpp:
```diff
@@ -25,6 +25,45 @@
 #include <vector>
 
+bool goto_symext::check_incremental(
+  expr2tc &expr,
```
expr can be const here: when calling assertion and assume, you can use gen_false_expr() directly.

lucasccordeiro replied:
Fixed.

mikhailramalho commented on src/goto-symex/symex_main.cpp:
```diff
@@ -54,6 +93,13 @@ void goto_symext::claim(const expr2tc &claim_expr, const std::string &msg)
+      // incremental verification has succeeded
+      return;
+  }
+
```
Shouldn't we add the result of the evaluation to the guards if the assertion evaluates to false?

lucasccordeiro replied:
Fixed.

lucasccordeiro committed:
add assertion to the target equation

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
add a new line

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
make expr const

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
moved the result of the evaluation to the guards

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

lucasccordeiro committed:
removed the checking for assume statements

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>

mikhailramalho reviewed (approved):
LGTM

lucasccordeiro committed:
pass the same expression to assertion

Signed-off-by: Lucas Cordeiro <lucasccordeiro@gmail.com>
